src/java/org/apache/cassandra/nodes/Nodes.java

```diff
@@ -0,0 +1,348 @@
+                PeerInfo updated = existingPeerInfo == null
+                                   ? update.apply(new PeerInfo().setPeerAddressAndPort(peer))
+                                   : update.apply(existingPeerInfo.duplicate()); // since we operate on mutable objects, we don't want to let the update function to operate on the live object
+
+                if (updated.getPeerAddressAndPort() == null)
+                    updated.setPeerAddressAndPort(peer);
+                else
+                    Preconditions.checkArgument(Objects.equals(updated.getPeerAddressAndPort(), peer));
+
+                save(existingPeerInfo, updated, blocking, force);
+                return updated;
+            }).duplicate(); // we don't want to return the live object thus we duplicate
+        }
+
+        public PeerInfo remove(InetAddressAndPort peer, boolean blocking)
+        {
+            AtomicReference<PeerInfo> removed = new AtomicReference<>();
+            internalMap.compute(peer, (key, existingPeerInfo) -> {
+                delete(peer, blocking);
+                removed.set(existingPeerInfo);
+                return null;
+            });
+            return removed.get();
+        }
+
+        /**
+         * Returns a peer information for a given address if the peer is known. Otherwise returns {@code null}.
+         */
+        public PeerInfo get(InetAddressAndPort peer)
+        {
+            PeerInfo info = internalMap.get(peer);
+            return info != null ? info.duplicate() : null;
+        }
+
+        /**
+         * Returns a stream of all known peers.
+         */
+        public Stream<PeerInfo> get()
+        {
+            return internalMap.values().map(PeerInfo::duplicate);
+        }
+
+        private void save(PeerInfo previousInfo, PeerInfo newInfo, boolean blocking, boolean force)
+        {
+            if (!force && Objects.equals(previousInfo, newInfo))
+                return;
+
+            Future<?> f = updateExecutor.submit(() -> {
+                nodesPersistence.savePeer(newInfo);
+                if (blocking)
+                    nodesPersistence.syncPeers();
+            });
+            if (blocking)
+                FBUtilities.waitOnFuture(f);
+        }
+
+        private Peers load()
+        {
+            nodesPersistence.loadPeers().forEach(info -> internalMap.compute(info.getPeerAddressAndPort(), (key, existingPeerInfo) -> info));
+            return this;
+        }
+
+        private void delete(InetAddressAndPort peer, boolean blocking)
+        {
+            Future<?> f = updateExecutor.submit(() -> {
+                nodesPersistence.deletePeer(peer);
+                if (blocking)
+                    nodesPersistence.syncPeers();
+            });
+
+            if (blocking)
+                FBUtilities.waitOnFuture(f);
+        }
+    }
+
+    public class Local
+    {
+        private final LoadingMap<InetAddressAndPort, LocalInfo> internalMap = new LoadingMap<>();
```

Comment:
I know how it looks. Unfortunately, I don't know any nice way to replace the backend of this with something like `AtomicReference` without creating lambdas each time we want to deal with the collection.

On the other hand, I would not bother that much with OSS because for local info they use a separate table with just one record - this exactly the same case.

Anyways, I've applied some improvements to the map itself (a bit unrelated) and make sure that in case of Local, the map has size 1.